src/MySqlConnector/MySqlClient/MySqlDataReader.cs

```diff
@@ -781,12 +830,32 @@ private enum State
 			NoMoreData,
 		}
 
+	    public async Task WaitDoneReadingAsync(CancellationToken cancellationToken=default(CancellationToken))
+	    {
+	        cancellationToken.ThrowIfCancellationRequested();
+	        await m_done_reading.WaitAsync(cancellationToken).ConfigureAwait(false);
+	        m_done_reading.Release();
+	    }
+
 		static readonly Task<bool> s_falseTask = Task.FromResult(false);
 		static readonly Task<bool> s_trueTask = Task.FromResult(true);
 
 		MySqlCommand m_command;
 		MySqlSession m_session;
-		State m_state;
+	    SemaphoreSlim m_done_reading = new SemaphoreSlim(1);
+	    State _m_state;
+	    State m_state
+	    {
+	        get { return _m_state; }
+	        set
+	        {
+	            if (value == State.None || value == State.NoMoreData)
+	            {
+	                m_done_reading.Release();
```

Comment:
I think that this library could raise an event when that happens. That's zero overhead for clients who don't need it, but Pomelo.EF could subscribe to the event and release its semaphore when that happens. Then you wouldn't need clients to dispose the reader, but simply keep reading until `Read(Async)` returns false.

This still seems like a programming error that shouldn't be automatically masked by Pomelo.EF, but I'm not familiar enough with EF to know if that's a standard pattern or not, and if not implementing it would inconvenience clients.